sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableInserter.java

```diff
@@ -273,6 +296,151 @@ public void insertAll(TableReference ref, List<TableRow> rowList,
+    loadConfig.setSourceUris(gcsUris);
+    loadConfig.setDestinationTable(ref);
+    loadConfig.setSchema(schema);
+    loadConfig.setWriteDisposition(writeDisposition.name());
+    loadConfig.setCreateDisposition(createDisposition.name());
+    loadConfig.setSourceFormat("NEWLINE_DELIMITED_JSON");
+
+    String projectId = ref.getProjectId();
+    for (int i = 0; i < MAX_RETRY_LOAD_JOBS; ++i) {
+      BackOff backoff = new AttemptBoundedExponentialBackOff(
+          MAX_LOAD_JOB_RPC_ATTEMPTS, INITIAL_LOAD_JOB_RPC_BACKOFF_MILLIS);
+      String retryingJobId = jobId + "-" + i;
+      insertLoadJob(retryingJobId, loadConfig, Sleeper.DEFAULT, backoff);
+      Status jobStatus = pollJobStatus(projectId, retryingJobId);
+      switch (jobStatus) {
+        case SUCCEEDED:
+          return;
+        case UNKNOWN:
+          throw new RuntimeException("Failed to poll the load job status.");
+        case FAILED:
+          continue;
+        default:
+          throw new IllegalStateException("Unexpected job status: " + jobStatus);
+      }
+    }
+    throw new RuntimeException(
+        "Failed to create the load job, reached max retries: " + MAX_RETRY_LOAD_JOBS);
+  }
+
+  @VisibleForTesting
+  void insertLoadJob(
+      String jobId,
+      JobConfigurationLoad loadConfig,
+      Sleeper sleeper,
+      BackOff backoff)
+      throws InterruptedException, IOException {
+    TableReference ref = loadConfig.getDestinationTable();
+    String projectId = ref.getProjectId();
+
+    Job job = new Job();
+    JobReference jobRef = new JobReference();
+    jobRef.setProjectId(projectId);
+    jobRef.setJobId(jobId);
+    job.setJobReference(jobRef);
+    JobConfiguration config = new JobConfiguration();
+    config.setLoad(loadConfig);
+    job.setConfiguration(config);
+
+    Exception lastException = null;
+    do {
+      try {
+        client.jobs().insert(projectId, job).execute();
+        return; // SUCCEEDED
+      } catch (GoogleJsonResponseException e) {
+        if (errorExtractor.itemAlreadyExists(e)) {
+          return; // SUCCEEDED
+        }
+        // ignore and retry
+        LOG.warn("Ignore the error and retry inserting the job.", e);
+        lastException = e;
+      } catch (IOException e) {
+        // ignore and retry
+        LOG.warn("Ignore the error and retry inserting the job.", e);
+        lastException = e;
+      }
+    } while (nextBackOff(sleeper, backoff));
+    throw new IOException(
+        String.format(
+            "Unable to insert job: %s, aborting after %d retries.",
+            jobId, MAX_LOAD_JOB_RPC_ATTEMPTS),
+        lastException);
+  }
+
+  @VisibleForTesting
+  enum Status {
+    SUCCEEDED,
+    FAILED,
+    UNKNOWN,
+  }
+
+  private Status pollJobStatus(String projectId, String jobId) throws InterruptedException {
+    BackOff backoff = new AttemptBoundedExponentialBackOff(
+        MAX_LOAD_JOB_POLL_RETRIES, INITIAL_LOAD_JOB_POLL_BACKOFF_MILLIS);
```

Comment:
as discussed, it looks like this will fail (returning `Status.UNKNOWN`) if the load job does not finish relatively fast. Load jobs can easily take an hour or more, so we need to be sure that we can support that.